internal/controllers/sharedvolume/controller.go

```diff
@@ -54,177 +61,149 @@ type Reconciler struct {
+			return false
+		},
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&v1.Volume{}).
+		WithEventFilter(filters).
+		WithLogger(r.log).
+		Complete(r)
 }
 
 // +kubebuilder:rbac:groups="",resources=services,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups="",resources=services/status,verbs=get;update;patch
 // +kubebuilder:rbac:groups="",resources=endpoints,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups="",resources=endpoints/status,verbs=get;update;patch
 
-// Start runs the main reconcile loop until the context is cancelled or there is
-// a fatal error.  It implements the controller-runtime Runnable interface so
-// that it can be controlled by controller manager.
-//
-// It polls the StorageOS api every apiPollInterval and ensures that the K8s
+// Reconcile the StorageOS volumes and ensures that the K8s
 // objects that are required for shared volumes are present.  It does not handle
 // deleting objects that are no longer required, this is done via
 // OwnerReferences.
-func (r *Reconciler) Start(ctx context.Context) error {
-	for {
-		tr := otel.Tracer("shared-volume")
-		ctx, span := tr.Start(ctx, "shared volume reconciler")
-
-		start := time.Now()
-
-		// Query shared volumes info.
-		volumes, err := r.api.ListSharedVolumes(ctx)
-		if err != nil {
-			span.RecordError(errors.Wrap(err, "failed to list shared volumes"))
-			r.log.Error(err, "failed to list shared volumes")
-			r.apiReset <- struct{}{}
-		}
-		span.SetAttributes(label.Int("volumes", len(volumes)))
-
-		for _, vol := range volumes {
-			log := r.log.WithValues("svc", vol.ServiceName, "pvc", vol.PVCName, "namespace", vol.Namespace)
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	start := time.Now()
 
-			// New tracing span for each volume.
-			ctx, span := tr.Start(ctx, "reconcile shared volume")
-			span.SetAttributes(label.String("pvc", vol.PVCName))
-			span.SetAttributes(label.String("namespace", vol.Namespace))
+	vol := &v1.Volume{}
 
-			observeErr := func(err error, msg string) {
-				span.RecordError(errors.Wrap(err, msg))
-				log.Error(err, msg)
-			}
+	err := r.Client.Get(ctx, req.NamespacedName, vol)
+	if apierrors.IsNotFound(err) {
+		return ctrl.Result{}, nil
+	} else if err != nil {
+		return ctrl.Result{}, err
+	}
 
-			// Fetch volume from cache. If the cached entry is the same then
-			// skip the k8s api requests to verify until the cache entry
-			// expires.
-			obj, found := r.volumes.Get(vol.ID)
-			if found {
-				cachedVol, ok := obj.(*storageos.SharedVolume)
-				if !ok {
-					log.Error(ErrCastCache, "failed to cast cache object to SharedVolume type", "object", obj)
-					span.RecordError(err)
-					span.End()
-					continue
-				}
-				if cachedVol.IsEqual(vol) {
-					span.AddEvent("no update needed")
-					span.End()
-					continue
-				}
-			}
+	sv := storageos.SharedVolume{Volume: vol}
```

Comment:
I'm wondering if you should use a constructor here - to be a valid SharedVolume it needs a few things (nfs endpoint, pvcName, etc.)  It might be better to validate in the constructor and return an error early?

Reply:
For SharedVolumes, we require certain labels to be set.  The constructor should validate and return an error if the volume object isn't compatible.